### PR TITLE
Make whitespace trimming in block selection configurable

### PR DESCRIPTION
**Pull request by Don-Vito**
## PR Checklist
* [x] Closes https://github.com/microsoft/terminal/issues/9706
* [x] CLA signed.
* [ ] Tests added/passed
* [x] Documentation updated here: https://github.com/MicrosoftDocs/terminal/pull/313
* [x] Schema updated.
* [ ] I've discussed this with core contributors already. 

## Detailed Description of the Pull Request / Additional comments
Added global flag named `trimBlockSelection` set to `false` by default.
The setting was added to Interactions menu of the SUI.

### CODE REVIEW

**Comment by Don-Vito** [comment]
image

**Review by miniksa — APPROVED** [review]
Looks good to me but...

> Documentation updated - not yet

We definitely need to file the doc bug or make sure @cinnamon-msft is ready for this.

**Comment by Don-Vito** [comment]
>  to file the doc bug or m

Thanks!. Issuing it now - waited for initial approval.

**Comment by Don-Vito** [comment]
> 
> 
> Looks good to me but...
> 
> > Documentation updated - not yet
> 
> We definitely need to file the doc bug or make sure @cinnamon-msft is ready for this.

The documentation was added here: https://github.com/MicrosoftDocs/terminal/pull/313

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 1]
```diff
@@ -209,6 +209,10 @@
     <value>Automatically copy selection to clipboard</value>
     <comment>Header for a control to toggle whether selected text should be copied to the clipboard automatically, or not.</comment>
   </data>
+  <data name="Globals_TrimBlockSelection.Header" xml:space="preserve">
+    <value>Remove trailing white-spaces when copying text in rectangular selection to clipboard</value>
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:213]: I'm ok with this. But @cinnamon-msft might want to take a look here haha

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 2]
```diff
@@ -255,15 +255,14 @@ const TextBuffer::TextAndColor Terminal::RetrieveSelectedTextFromBuffer(bool sin
...
-    return _buffer->GetText(!singleLine || _blockSelection,
-                            !singleLine && !_blockSelection,
-                            selectionRects,
-                            GetAttributeColors,
-                            _blockSelection);
+    // GH#9706: Trimming of trailing white-spaces in block selection is configurable.
+    const auto includeCRLF = !singleLine || _blockSelection;
+    const auto trimTrailingWhitespace = !singleLine && (!_blockSelection || _trimBlockSelection);
+    const auto formatWrappedRows = _blockSelection;
+    return _buffer->GetText(includeCRLF, trimTrailingWhitespace, selectionRects, GetAttributeColors, formatWrappedRows);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalCore/TerminalSelection.cpp:265]: THIS LOOKS SO MUCH BETTER. THANK YOU

**Review by carlos-zamora — APPROVED** [review]
Looks great! Thanks!

**Review by DHowett — CHANGES_REQUESTED** [review]
Blocking so that we have space for a discussion about where the setting belongs (which level), naming, phrasing, etc.

2 signoffs is dangerously close to "it merges without us agreeing on stuff" ;P

**Comment by Don-Vito** [comment]
@DHowett - requested review just to ensure that the bot removes "Needs-Author-Feedback".

**Comment by carlos-zamora** [comment]
I'll kick off the discussion then 😊

> where the setting belongs (which level)

I think this makes sense as a global. It's next to `copyOnSelect`, so the user can expect a consistent selection interaction across all of their sessions.

> naming

`trimBlockSelection`: `<bool>`

I actually _really_ like this name. It's short. It's clear that it _only_ impacts block selection.

My only complaint could be that the trimming only happens when we copy, not when we select. But I think that's me just grasping at straws to find something wrong with it haha. And I honestly think people won't think of the setting in that way; docs, the schema description, and the SUI header should be more than enough to remove that ambiguity.

> phrasing, etc.

"Remove trailing white-spaces when copying text in rectangular selection to clipboard"

Maybe change "rectangular" to "block". I'm mildly irritated by how long it is, but I'd rather see it be descriptive than not. And I think this is ok to be "negative", because "Keep trailing white-spaces..." isn't clear that the alternative is to remove them. But yeah, definitely want more input on this one.

### Miscellaneous design thoughts

We _could_ change this setting to `trimCopiedText`, but I _really_ don't like that because I like that we're already handling trimming text pretty smart. Line selection doesn't trim, whereas block selection is up to the user. From [this comment](https://github.com/microsoft/terminal/issues/9706#issuecomment-814032515), it seems like there's already consensus on block selection being configurable in this way so this seems like a solid win for me.

**Comment by miniksa** [comment]
I concur with all of @carlos-zamora's points and don't have anything further to add.

**Comment by Don-Vito** [comment]
It's been a week. So bump.

**Review by zadjii-msft — APPROVED** [review]
Yea this is super straightforward, I have no problems with this. /discussion

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 1 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:213]: @carlos-zamora how do you feel about this

```suggestion
    <value>Remove trailing white-space in rectangular selection</value>
```

I am okay with the negative boolean. It _enables a destructive operation_.

**Review by DHowett — APPROVED** [review]
Shopping around the new wording, then we'll merge. Thanks for the patience!

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw (hunk 1 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:213]: I like that a lot more. This is one of those cases where I'd prefer it to be short but ambiguous. Thanks!

**Comment by 89z** [comment]
https://github.com/microsoft/terminal/issues/12536

**Pull request merged**